pkg/kv/kvserver/concurrency/lock_table_waiter.go

```diff
@@ -648,6 +666,68 @@ func (c *txnCache) insertFrontLocked(txn *roachpb.Transaction) {
 	c.txns[0] = txn
 }
 
+// contentionEventHelper tracks and emits ContentionEvents.
+type contentionEventHelper struct {
+	sp      *tracing.Span
+	ev      *roachpb.ContentionEvent
+	tBegin  time.Time
+	onEvent func(event *roachpb.ContentionEvent) // may be nil
+}
+
+// emit emits the open contention event, if any.
+func (h *contentionEventHelper) emit() {
+	if h.ev == nil {
+		return
+	}
+	h.ev.Duration = timeutil.Since(h.tBegin)
+	if h.onEvent != nil {
+		// NB: this is intentionally above the call to LogStructured so that
+		// this interceptor gets to mutate the event (used for test determinism).
+		h.onEvent(h.ev)
+	}
+	h.sp.LogStructured(h.ev)
+	h.ev = nil
+}
+
+// emitAndInit compares the waitingState's active txn (if any) against the open
+// ContentionEvent (if any). If the they match, we are continuing to handle the
+// same event and no action is taken. If they differ, the open event (if any) is
```

Comment:
It would be a bit easier to read the implementation below if we had an explicit early return for this case.